flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/RestoreStreamTaskTest.java

```diff
@@ -0,0 +1,325 @@
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			new OperatorID(42L, 42L),
+			new CounterOperator(),
+			new OperatorID(44L, 44L),
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				new OperatorID(4242L, 4242L),
+				new RestoreCounterOperator(),
+				new OperatorID(44L, 44L),
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(1, RestoreCounterOperator.RESTORE_COUNTER.get());
+		}
+		finally {
+			RestoreCounterOperator.RESTORE_COUNTER.getAndSet(0);
+		}
+	}
+
+	@Test
+	public void testRestoreAfterScaleUp() throws Exception {
+		OperatorID headOperatorID = new OperatorID(42L, 42L);
+		OperatorID tailOperatorID = new OperatorID(44L, 44L);
+
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			headOperatorID,
+			new CounterOperator(),
+			tailOperatorID,
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		// test empty state in case of scale up
+		OperatorSubtaskState emptyHeadOperatorState = StateAssignmentOperation.operatorSubtaskStateFrom(
+			new OperatorInstanceID(0, headOperatorID),
+			Collections.emptyMap(),
+			Collections.emptyMap(),
+			Collections.emptyMap(),
+			Collections.emptyMap());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+		stateHandles.putSubtaskStateByOperatorID(headOperatorID, emptyHeadOperatorState);
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				headOperatorID,
+				new RestoreCounterOperator(),
+				tailOperatorID,
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(2, RestoreCounterOperator.RESTORE_COUNTER.get());
+		}
+		finally {
+			RestoreCounterOperator.RESTORE_COUNTER.getAndSet(0);
+		}
+	}
+
+	@Test
+	public void testRestoreWithoutState() throws Exception {
+		OperatorID headOperatorID = new OperatorID(42L, 42L);
+		OperatorID tailOperatorID = new OperatorID(44L, 44L);
+
+		AcknowledgeStreamMockEnvironment environment1 = processRecords(
+			headOperatorID,
+			new StatelessRestoreCounterOperator(),
+			tailOperatorID,
+			new CounterOperator(),
+			Optional.empty());
+
+		assertEquals(2, environment1.getCheckpointStateHandles().getSubtaskStateMappings().size());
+
+		TaskStateSnapshot stateHandles = environment1.getCheckpointStateHandles();
+
+		try {
+			AcknowledgeStreamMockEnvironment environment2 = processRecords(
+				headOperatorID,
+				new StatelessRestoreCounterOperator(),
+				tailOperatorID,
+				new RestoreCounterOperator(),
+				Optional.of(stateHandles));
+
+			assertEquals(1, StatelessRestoreCounterOperator.RESTORE_COUNTER.get());
```

Comment:
In theory, this could mean the exact opposite from the expected result because we don't know for which operator the `isRestored` was counted as `true` and for which as `false`.